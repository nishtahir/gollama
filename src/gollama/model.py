import json


class OllamaManifest:
    def __init__(self, json: dict):
        self.json = json

    def config(self) -> dict:
        return self.json["config"]

    def layers(self) -> list[dict]:
        return self.json["layers"]

    def blobs(self) -> list[str]:
        blobs = []
        blobs.append(self.config()["digest"])

        for layer in self.layers():
            blobs.append(layer["digest"])
        return blobs

    @staticmethod
    def from_path(path: str) -> "OllamaManifest":
        with open(path, "r") as f:
            return OllamaManifest(json.load(f))


class OllamaModel:
    def __init__(self, name: str, version: str):
        self.name = name
        self.version = version

    def __str__(self):
        return f"{self.name}:{self.version}"

    @staticmethod
    def from_name(model_name: str) -> "OllamaModel":
        if ":" not in model_name:
            model_name = f"{model_name}:latest"

        name, version = model_name.split(":")

        # we assume that model names with no slashes are
        # pulled from the ollama model registry
        if "/" not in name:
            name = f"registry.ollama.ai/library/{name}"

        return OllamaModel(name=name, version=version)

    @staticmethod
    def from_manifest_path(path: str) -> "OllamaModel":
        # models/manifests/registry.ollama.ai/library/smollm2/135m -> smollm2:135m
        name = path.split("/")[-2]
        version = path.split("/")[-1]
        return OllamaModel(name=name, version=version)
