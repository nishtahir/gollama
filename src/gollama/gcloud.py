import os

from google.cloud import storage
from tqdm import tqdm


def upload_blob(file_path: str, bucket_name: str, blob_name: str) -> None:
    storage_client = storage.Client()
    bucket = storage_client.bucket(bucket_name)
    blob = bucket.blob(blob_name)

    # take the file name from the blob name
    file_name = os.path.basename(blob_name)
    # remove the sha256- prefix
    if file_name.startswith("sha256-"):
        file_name = file_name.split("-")[1][:12]

    with open(file_path, "rb") as in_file:
        total_bytes = os.fstat(in_file.fileno()).st_size
        with tqdm.wrapattr(
            in_file,
            "read",
            total=total_bytes,
            miniters=1,
            desc=f"pushing {file_name}...",
        ) as file_obj:
            blob.upload_from_file(file_obj)


def download_blob(bucket_name: str, blob_name: str, file_path: str) -> None:
    storage_client = storage.Client()
    bucket = storage_client.bucket(bucket_name)
    blob = bucket.get_blob(blob_name)
    if blob is None:
        raise ValueError(f"Blob {blob_name} not found in bucket {bucket_name}")

    size = blob.size
    # take the file name from the blob name
    file_name = os.path.basename(blob_name)
    # remove the sha256- prefix
    if file_name.startswith("sha256-"):
        file_name = file_name.split("-")[1][:12]

    os.makedirs(os.path.dirname(file_path), exist_ok=True)
    with open(file_path, "wb") as out_file:
        with tqdm.wrapattr(
            out_file,
            "write",
            total=size,
            miniters=1,
            desc=f"pulling {file_name}...",
        ) as file_obj:
            blob.download_to_file(file_obj)


def list_blobs(bucket_name: str, prefix: str) -> list[str]:
    storage_client = storage.Client()
    bucket = storage_client.bucket(bucket_name)
    blobs = bucket.list_blobs(prefix=prefix)
    return [blob.name for blob in blobs]


def blob_exists(bucket_name: str, blob_name: str) -> bool:
    storage_client = storage.Client()
    bucket = storage_client.bucket(bucket_name)
    blob = bucket.blob(blob_name)
    return blob.exists()
