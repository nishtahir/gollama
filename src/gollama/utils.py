import os

__PROJECT_ROOT__ = os.path.abspath(os.path.join(os.path.dirname(__file__), "../.."))
