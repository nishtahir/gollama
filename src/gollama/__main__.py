import os
from typing import Annotated

import typer
from typer import Typer

import gollama.gcloud as gcloud
from gollama.model import OllamaManifest, OllamaModel

app = Typer(no_args_is_help=True)


@app.command("push", help="Push a model to a storage bucket")
def push(
    *,
    model: Annotated[str, typer.Argument(help="Model to push")],
    ollama_home: Annotated[str, typer.Option(envvar="OLLAMA_HOME")] = "~/.ollama",
    bucket_name: Annotated[str, typer.Option(envvar="GCS_BUCKET_NAME")],
):
    ollama_home = os.path.expanduser(ollama_home)
    manifests_dir = os.path.join("models", "manifests")
    blobs_dir = os.path.join("models", "blobs")

    ollama_model = OllamaModel.from_name(model)
    manifest_path = os.path.join(manifests_dir, ollama_model.name, ollama_model.version)

    abs_manifest_path = os.path.join(ollama_home, manifest_path)
    manifest = OllamaManifest.from_path(abs_manifest_path)
    # upload the manifest

    print("pushing manifest")
    gcloud.upload_blob(abs_manifest_path, bucket_name, manifest_path)

    blobs = manifest.blobs()

    for blob in blobs:
        blob_name = blob.replace(":", "-")
        blob_path = os.path.join(blobs_dir, blob_name)
        abs_blob_path = os.path.join(ollama_home, blob_path)
        gcloud.upload_blob(abs_blob_path, bucket_name, blob_path)


@app.command("list", help="List models in a storage bucket")
def list(
    *,
    bucket_name: Annotated[str, typer.Option(envvar="GCS_BUCKET_NAME")],
):
    manifests_dir = os.path.join("models", "manifests")
    blobs = gcloud.list_blobs(bucket_name, manifests_dir)
    for manifest in blobs:
        model = OllamaModel.from_manifest_path(manifest)
        print(model)


@app.command("pull", help="Pull a model from a storage bucket")
def pull(
    *,
    model: Annotated[str, typer.Argument(help="Model to pull")],
    ollama_home: Annotated[str, typer.Option(envvar="OLLAMA_HOME")] = "~/.ollama",
    bucket_name: Annotated[str, typer.Option(envvar="GCS_BUCKET_NAME")],
):
    ollama_home = os.path.expanduser(ollama_home)
    manifests_dir = os.path.join("models", "manifests")
    blobs_dir = os.path.join("models", "blobs")

    # check if the model exists in the bucket
    ollama_model = OllamaModel.from_name(model)
    manifest_path = os.path.join(manifests_dir, ollama_model.name, ollama_model.version)
    abs_manifest_path = os.path.join(ollama_home, manifest_path)

    # download the manifest
    print("pulling manifest")
    gcloud.download_blob(bucket_name, manifest_path, abs_manifest_path)
    manifest = OllamaManifest.from_path(abs_manifest_path)

    # download the blobs
    blobs = manifest.blobs()
    for blob in blobs:
        blob_name = blob.replace(":", "-")
        blob_path = os.path.join(blobs_dir, blob_name)
        abs_blob_path = os.path.join(ollama_home, blob_path)
        gcloud.download_blob(bucket_name, blob_path, abs_blob_path)


if __name__ == "__main__":
    app()
