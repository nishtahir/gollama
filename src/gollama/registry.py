import json
from datetime import datetime

from google.cloud.storage import Blob as GcloudBlob

import gollama.gcloud as gcloud


class Manifest:
    def __init__(self, bucket_name: str, blob: GcloudBlob, text: str):
        self.bucket_name = bucket_name
        self.blob = blob
        self.content = json.loads(text)

    @property
    def name(self) -> str:
        path: str = self.blob.name  # type: ignore
        path = path.removeprefix("models/manifests/")
        path = path.removeprefix("registry.ollama.ai/")
        path = path.removeprefix("library/")

        name, version = path.rsplit("/", 1)
        return f"{name}:{version}"

    @property
    def details(self) -> dict:
        return self.content

    @property
    def modified(self) -> datetime:
        # remove timezone info for timeago
        return self.blob.updated.replace(tzinfo=None)  # type: ignore

    @property
    def size(self) -> int:
        return sum(layer["size"] for layer in self.content["layers"])

    def __str__(self):
        return f"Manifest(blob={self.blob.name})"

    def __repr__(self):
        return self.__str__()


class Blob:
    def __init__(self, blob: GcloudBlob):
        self.blob = blob

    def __str__(self):
        return f"Blob(blob={self.blob.name})"

    def __repr__(self):
        return self.__str__()


class GcloudRegistry:
    def __init__(self, manifests: list[Manifest], blobs: list[Blob]):
        self._manifests = manifests
        self._blobs = blobs

    @property
    def manifests(self) -> list[Manifest]:
        return self._manifests

    @property
    def blobs(self) -> list[Blob]:
        return self._blobs

    def has_blob(self, path: str) -> bool:
        return any(blob.blob.name == path for blob in self.blobs)

    def __str__(self):
        return f"GcloudRegistry(manifests={self.manifests}, blobs={self.blobs})"

    def __repr__(self):
        return self.__str__()


def create_registry(bucket_name: str) -> GcloudRegistry:
    manifest_blobs = gcloud.list_blobs(bucket_name, "models/manifests")
    contents = map(lambda blob: blob.download_as_text(), manifest_blobs)

    iterator = zip(manifest_blobs, contents)
    manifests = [Manifest(bucket_name, blob, content) for blob, content in iterator]

    blobs = [Blob(blob) for blob in gcloud.list_blobs(bucket_name, "models/blobs")]
    return GcloudRegistry(manifests, blobs)
