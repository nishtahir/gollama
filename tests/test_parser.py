import os

from gollama.model import OllamaManifest, OllamaModel
from gollama.utils import __PROJECT_ROOT__


def test_parse_ollama_model_from_name(snapshot):
    names = [
        "gemma3",
        "gemma3:12b",
        "microsoft/phi-4:latest",
        "gemma3:4b",
        "hf.co/lmstudio-community/zeta-GGUF:Q4_K_M",
        "llama3.1:8b",
        "qwen2.5-coder:1.5b-base",
        "nomic-embed-text:latest",
    ]

    models = list(map(OllamaModel.from_name, names))
    snapshot.assert_match(models)


def test_parse_ollama_manifest_from_path(snapshot):
    manifest = OllamaManifest.from_path(
        os.path.join(
            __PROJECT_ROOT__,
            "tests/fixtures/microsoft-phi-4-latest",
        )
    )
    snapshot.assert_match(manifest)


def test_parse_ollama_manifest_blobs(snapshot):
    manifest = OllamaManifest.from_path(
        os.path.join(
            __PROJECT_ROOT__,
            "tests/fixtures/microsoft-phi-4-latest",
        )
    )
    blobs = manifest.blobs()
    snapshot.assert_match(blobs)
