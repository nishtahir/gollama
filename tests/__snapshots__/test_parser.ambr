# serializer version: 1
# name: test_parse_ollama_manifest_blobs
  list([
    'sha256:d8e5f6443e0ba6195753078b44cf8dfafe46c89109208ca28835591afdc8c36f',
    'sha256:de3ceb05eea4beeb69e7819c35003039c4439117c4cf44e31d736a2806662bbb',
    'sha256:e3b2721cec511e3a627ae0993753abbf5c72cdf4cfcdbc8f5c33ce15c4c4e04b',
    'sha256:f02dd72bb2423204352eabc5637b44d79d17f109fdb510a7c51455892aa2d216',
  ])
# ---
# name: test_parse_ollama_manifest_from_path
  OllamaManifest(
    json=dict({
      'config': dict({
        'digest': 'sha256:d8e5f6443e0ba6195753078b44cf8dfafe46c89109208ca28835591afdc8c36f',
        'mediaType': 'application/vnd.docker.container.image.v1+json',
        'size': 412,
      }),
      'layers': list([
        dict({
          'digest': 'sha256:de3ceb05eea4beeb69e7819c35003039c4439117c4cf44e31d736a2806662bbb',
          'mediaType': 'application/vnd.ollama.image.model',
          'size': 9053114240,
        }),
        dict({
          'digest': 'sha256:e3b2721cec511e3a627ae0993753abbf5c72cdf4cfcdbc8f5c33ce15c4c4e04b',
          'mediaType': 'application/vnd.ollama.image.template',
          'size': 102,
        }),
        dict({
          'digest': 'sha256:f02dd72bb2423204352eabc5637b44d79d17f109fdb510a7c51455892aa2d216',
          'mediaType': 'application/vnd.ollama.image.params',
          'size': 59,
        }),
      ]),
      'mediaType': 'application/vnd.docker.distribution.manifest.v2+json',
      'schemaVersion': 2,
    }),
  )
# ---
# name: test_parse_ollama_model_from_name
  list([
    OllamaModel(
      name='registry.ollama.ai/library/gemma3',
      version='latest',
    ),
    OllamaModel(
      name='registry.ollama.ai/library/gemma3',
      version='12b',
    ),
    OllamaModel(
      name='microsoft/phi-4',
      version='latest',
    ),
    OllamaModel(
      name='registry.ollama.ai/library/gemma3',
      version='4b',
    ),
    OllamaModel(
      name='hf.co/lmstudio-community/zeta-GGUF',
      version='Q4_K_M',
    ),
    OllamaModel(
      name='registry.ollama.ai/library/llama3.1',
      version='8b',
    ),
    OllamaModel(
      name='registry.ollama.ai/library/qwen2.5-coder',
      version='1.5b-base',
    ),
    OllamaModel(
      name='registry.ollama.ai/library/nomic-embed-text',
      version='latest',
    ),
  ])
# ---
